pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -146,49 +111,74 @@ impl BlobWriter for EphemeralFile {
                     off: (ephemeral_file.size % PAGE_SZ as u64) as usize,
-                    memo_page_guard: MemoizedPageWriteGuard {
-                        guard: ephemeral_file.get_buf_for_write(blknum)?,
-                        blknum,
-                    },
                     ephemeral_file,
                 })
             }
             #[inline(always)]
             fn push_bytes(&mut self, src: &[u8]) -> Result<(), io::Error> {
-                // `src_remaining` is the remaining bytes to be written
                 let mut src_remaining = src;
                 while !src_remaining.is_empty() {
-                    let page = if self.memo_page_guard.blknum == self.blknum {
-                        &mut self.memo_page_guard.guard
-                    } else {
-                        self.memo_page_guard.guard =
-                            self.ephemeral_file.get_buf_for_write(self.blknum)?;
-                        self.memo_page_guard.blknum = self.blknum;
-                        &mut self.memo_page_guard.guard
-                    };
-                    let dst_remaining = &mut page[self.off..];
+                    let dst_remaining = &mut self.ephemeral_file.mutable_tail[self.off..];
                     let n = min(dst_remaining.len(), src_remaining.len());
                     dst_remaining[..n].copy_from_slice(&src_remaining[..n]);
                     self.off += n;
                     src_remaining = &src_remaining[n..];
                     if self.off == PAGE_SZ {
-                        // This block is done, move to next one.
-                        self.blknum += 1;
-                        self.off = 0;
+                        match self.ephemeral_file.file.write_all_at(
+                            &self.ephemeral_file.mutable_tail,
+                            self.blknum as u64 * PAGE_SZ as u64,
+                        ) {
+                            Ok(_) => {
+                                // Pre-warm the page cache with what we just wrote.
+                                // This isn't necessary for coherency/correctness, but it's how we've always done it.
+                                let cache = page_cache::get();
+                                match cache.read_immutable_buf(
+                                    self.ephemeral_file.page_cache_file_id,
+                                    self.blknum,
+                                ) {
+                                    Err(e) => {
+                                        error!("ephemeral_file write_blob failed to get immutable buf to pre-warm page cache: {e:?}");
+                                        // fail gracefully, it's not the end of the world if we can't pre-warm the cache here
+                                    }
+                                    Ok(page_cache::ReadBufResult::Found(_guard)) => {
+                                        // This function takes &mut self, so, it shouldn't be possible to reach this point.
```

Comment:
So we have &mut EphmeralFile and it makes "Found" result impossible?

Reply:
Yeah, not clear enough from the comment?

Reply:
Is it because this block we wrote is new?

Reply:
> Is it because this block we wrote is new?

yes

Reply:
> And there's zero chance of anyone reading the file after we write it and before we read it, because of a Mutex somewhere giving validity to &mut EphmeralFile => &mut fd?

There's no fd underneath, it's an `Arc<VirtualFile>`.
But yeah, that `Arc<VirtualFile>` is exclusively owned by EphemeralFile.

Reply:
Hm, I think I can get rid of the Arc

Reply:
e0b8785f72f9dd9a588e203b556fc97f23eaadf2

Reply:
Exclusively owned Arc sounds interesting :) Ok, not the thing I was thinking of but e0b8785 seems like a simplification.

The change will not touch how many open files or anything like that, because VirtualFile still handles it underneath I assume.

Reply:
Fine to resolve if I am not completly wrong in my assumptions.

Reply:
> The change will not touch how many open files or anything like that, because VirtualFile still handles it underneath I assume.

correct